### PR TITLE
raftstore: delay replica read when applying snapshot

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

###  What have you changed?

When a peer is applying a snapshot, it will clean up the data of the region and ingest the snapshot file.
If a replica read happens between the two operations, the empty data will be returned.
This PR delays all replica reads when the peer is applying snapshot.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.

###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1076,7 +1076,10 @@ impl Peer {
         // problem if the leader applies fewer values than the follower, the follower read
         // could get a newer value, and after that, the leader may read a stale value,
         // which violates linearizability.
-        self.get_store().applied_index() >= read_index && !self.is_splitting() && !self.is_merging()
+        self.get_store().applied_index() >= read_index
+            && !self.is_splitting()
+            && !self.is_merging()
+            && !self.is_applying_snapshot()
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: Please add some comments to explain why we need to check `is_apply_snapshot`?

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -1451,7 +1455,10 @@ impl Peer {
                         "peer_id" => self.peer.get_id(),
                     );
                     for (req, cb) in read.cmds.drain(..) {
-                        if req.get_header().get_replica_read() {
+                        // We should check epoch since the range could be changed
+                        if req.get_header().get_replica_read()
+                            && util::check_region_epoch(&req, self.region(), false).is_ok()
+                        {
                             cb.invoke_read(self.handle_read(ctx, req, true, read.read_index));
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: `handle_read` checks epoch too.

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1471,14 +1478,14 @@ impl Peer {
...
-        if !self.is_leader() && !ready.read_states().is_empty() {
+        if !self.is_leader() {
             // NOTE: there could still be some read requests following, which will be cleared in
             // `clear_uncommitted` later.
             for state in ready.read_states() {
                 self.pending_reads
                     .advance(state.request_ctx.as_slice(), state.index);
-                self.post_pending_read_index_on_replica(ctx);
             }
+            self.post_pending_read_index_on_replica(ctx);
```
Comment (Connor1996) [inline src/raftstore/store/peer.rs]: why calling it even `read_states` is empty?

Comment (5kbpers) [reply src/raftstore/store/peer.rs]: For handling the delayed requests.

**Pull request merged**